### PR TITLE
optimize dispersive pulse

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Remove duplicated calculations by storing intermediate results.
Add getter for the focus and origin in the base pulse functor. Make laser axis unit vector public.

I am free to change the name of the helper class, I do not found a better name. 
I do not used the helper variables in the equations directly to reduce the changelog and simplify the review.

The second commit is fixing an invalid memory access.

- [x] rebase against #4927
- [x] rebase against #4947

### CODE REVIEW

File: include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp [hunk 1]
```diff
@@ -79,6 +79,63 @@ namespace picongpu
...
+                            float_X omega;
+
+                            /**
+                             * @param totalCellIdx cell index in the total domain (including all moving window slides)
+                             * @param Omega frequency for which the E-value is calculated
+                             */
+                            HDINLINE
+                            LaserParams(floatD_X const& totalCellIdx, float_X const Omega) : omega(Omega)
+                            {
+                                internPos = this->getInternalCoordinates(totalCellIdx);
```
Comment (steindev) [inline include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp]: `this->` refers to the LaserParams class now, which does not have a `getInternalCoordinates` function

Comment (psychocoderHPC) [reply include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp]: fixed

**Comment by PrometheusPi** [comment]
The code fails with the following error:
```
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/incidentField/profiles/DispersivePulse.hpp:316:62: error: 'LaserParams' does not refer to a value
                                auto laserParams = Unitless::LaserParams(static_cast<Base>(*this), totalCellIdx, Omk);
                                                   ~~~~~~~~~~^
```

**Comment by psychocoderHPC** [comment]
> The code fails with the following error:
> 
> ```
> /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/incidentField/profiles/DispersivePulse.hpp:316:62: error: 'LaserParams' does not refer to a value
>                                 auto laserParams = Unitless::LaserParams(static_cast<Base>(*this), totalCellIdx, Omk);
>                                                    ~~~~~~~~~~^
> ```

stupid compiler, all worked in my tests. I hope I fixed it now.

**Comment by psychocoderHPC** [comment]
@steindev @PrometheusPi The second commit is fixing laser bug for the B field in 2-dimensional simulations.

File: include/picongpu/fields/incidentField/Functors.hpp [hunk 2]
```diff
@@ -336,7 +376,7 @@ namespace picongpu
                         auto const& subGrid = Environment<simDim>::get().SubGrid();
                         auto const globalDomainCells = subGrid.getGlobalDomain().size;
                         auto const direction = getDirection();
-                        auto const focus = getFocus();
+                        auto const focus = projectFocus();
```
Comment (steindev) [inline include/picongpu/fields/incidentField/Functors.hpp:379]: For future refactoring: This shadows the struct's member variable. Yet it is set to the same value...

**Comment by psychocoderHPC** [comment]
I opened a PR for the fix only and will rebase this PR asap after merging #4927

**Comment by steindev** [comment]
Replaced by #4932.

**Comment by steindev** [comment]
Closed wrong PR :see_no_evil:

**Comment by psychocoderHPC** [comment]
PR can be reviewed

**Comment by ikbuibui** [comment]
@psychocoderHPC PR can now be rebased against #4947

**Comment by psychocoderHPC** [comment]
I will close this PR with the optimization of the equation in https://github.com/ComputationalRadiationPhysics/picongpu/pull/4947 this PR will only save a small amount of operations. A refactoring where we add focus and center as arguments to `amp()` and `phi()` will be much more readable compared to the helper class with common values.

**Pull request closed without merging**